test/files/run/t7187.scala

```diff
@@ -0,0 +1,6 @@
+object Test extends App {
+  def foo(): () => String = () => "hi there"
+  val f: () => Any = foo
```

Comment:
Note that in 2.13 (`-Xsource:2.13`) this compiles without a warning, `foo` is applied to the empty parameter list.

In 2.12, eta-expansion comes in before, and it compiles with a deprecation warning:
```
➜  sandbox git:(syntacticNamedArgs-213) scalac Test.scala -deprecation
Test.scala:3: warning: Eta-expansion of zero-argument method values is deprecated. Did you intend to write Test.this.foo()?
  val f: () => Any = foo
                     ^
one warning found
```

So the program compiles in both versions, but differently.